File: bin/bubblesort/bubblesort.js
const fs = require('fs');
const path = require('path');

// Definindo path para o arquivo a ser lido e onde salvar o novo arquivo

const inputPathArquivo = path.join(__dirname, 'arq.txt'); 
const outputPathArquivo = path.join(__dirname, 'arquivo_ordenado.txt'); 

// Array para manipular os numeros

let arrayNumeros = [];

// Função para printar o uso de memoria e o tempo de execução
function printPerformance(startTime, startMemory) {
    const endTime = process.hrtime(startTime);
    const endMemory = process.memoryUsage();
    
    console.log('Tempo de execução: %ds %dms', endTime[0], endTime[1] / 1e6);
    console.log('Memória usada:');
    console.log('  Heap Used: %d KB', endMemory.heapUsed / 1024);
}

// Função para ler o arquivo e adicionar no Array criado a cima

function lerArquivo(filePath, callback) {
    const startTime = process.hrtime();
    const startMemory = process.memoryUsage();

    fs.readFile(filePath, 'utf8', (err, data) => {
        // Caso de erro
        if (err) {
            console.error('Erro ao ler o arquivo:', err);
            return;
        }
        
        // Separando os numeros pela quebra de linha
        const linhas = data.split('\n');
        // Criando um mapa com os numeros
        const numeros = linhas.map(linhas => parseFloat(linhas.trim())).filter(num => !isNaN(num));
        // Adicionando os numeros ao array
        arrayNumeros = arrayNumeros.concat(numeros); 
        // Usando a função para o print da performance
        printPerformance(startTime, startMemory);
        // Usando a função de callback
        callback();
    });
}

// Função de Bubble Sort
function bubbleSort(arr) {
    let len = arr.length;
    for (let i = 0; i < len; i++) {
        for (let j = 0; j < len - 1 - i; j++) {
            if (arr[j] > arr[j + 1]) {
                let temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
            }
        }
    }
}

// Função para gravar o array ordenado em um novo arquivo
function escreverArquivo(filePath, numerosArray, callback) {
    const data = numerosArray.join('\n') + '\n'; // Formata os números como linhas separadas

    const startTime = process.hrtime();
    const startMemory = process.memoryUsage();

    fs.writeFile(filePath, data, 'utf8', (err) => {
        if (err) {
            console.error('Erro ao gravar o arquivo:', err);
            return;
        }
        console.log('Arquivo gravado com sucesso:', filePath);

        // Mensurar o desempenho após a gravação do arquivo
        printPerformance(startTime, startMemory);

        callback();
    });
}

// Função de callback para usar o array após a leitura, ordenação e gravação
function ordenarESalvar() {
    bubbleSort(arrayNumeros);
    escreverArquivo(outputPathArquivo, arrayNumeros, () => {
        console.log('Processo concluído.');
    });
}

// Chamando a função para iniciar o programa
lerArquivo(inputPathArquivo, ordenarESalvar);

File: src/bubblesort/bubblesort.java
package bubblesort;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class bubblesort {
    // Caminhos dos arquivos
    private static final String inputPathArquivo = "arq.txt";
    private static final String outputPathArquivo = "arquivo_ordenadoJava.txt";

    // Lista para armazenar os números
    private static List<Double> numerosList = new ArrayList<>();

    // Main para rodar tudo
    public static void main(String[] args) {
        long startTime = System.currentTimeMillis();
        Runtime runtime = Runtime.getRuntime();
        long startMemory = runtime.totalMemory() - runtime.freeMemory();

        lerNumeros(inputPathArquivo);
        
        bubbleSort(numerosList);

        escreverArquivo(outputPathArquivo);

        long endTime = System.currentTimeMillis();
        long endMemory = runtime.totalMemory() - runtime.freeMemory();

        printPerformance(startTime, endTime, startMemory, endMemory);

        System.out.println("Processo concluído.");
    }

    // Função para ler os numeros do arquivo e colocar no ArrayList
    private static void lerNumeros(String filePath) {
        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
            String linha;
            while ((linha = br.readLine()) != null) {
                try {
                    double numeros = Double.parseDouble(linha.trim());
                    numerosList.add(numeros);
                } catch (NumberFormatException e) {
                    // Ignorar linhas que não são números
                }
            }
        } catch (IOException e) {
            System.err.println("Erro ao ler o arquivo: " + e.getMessage());
        }
    }

    // Função para ordenar o ArrayList
    private static void bubbleSort(List<Double> list) {
        int n = list.size();
        for (int i = 0; i < n - 1; i++) {
            for (int j = 0; j < n - i - 1; j++) {
                if (list.get(j) > list.get(j + 1)) {
                    Collections.swap(list, j, j + 1);
                }
            }
        }
    }

    // Função para escrever no arquivo novo
    private static void escreverArquivo(String filePath) {
        try (FileWriter fw = new FileWriter(filePath)) {
            for (Double numero : numerosList) {
                fw.write(numero + "\n");
            }
        } catch (IOException e) {
            System.err.println("Erro ao gravar o arquivo: " + e.getMessage());
        }
    }

    // Função para printar o tempo de execução e a memoria usada
    private static void printPerformance(long startTime, long endTime, long startMemory, long endMemory) {
        long elapsedTime = endTime - startTime;
        long usedMemory = (endMemory - startMemory) / 1024; // em KB

        System.out.println("Tempo de execução: " + (elapsedTime / 1000) + "s " + (elapsedTime % 1000) + "ms");
        System.out.println("Memória usada: " + usedMemory + " KB");
    }
    
}